include/kernels/staggered_kd_apply_xinv_kernel.cuh

```diff
@@ -1,219 +1,110 @@
-    const fineColorSpinor in; /** Input staggered spinor field */
-    const xInvGauge xInv;     /** Kahler-Dirac inverse field */
-
-    int_fastdiv x_size[QUDA_MAX_DIM];   /** Dimensions of fine grid */
-    int_fastdiv xc_size[QUDA_MAX_DIM];  /** Dimensions of coarse grid */
+    static constexpr bool dagger = dagger_;
 
-    const int fineVolumeCB;
-    const int_fastdiv coarseVolumeCB;   /** Coarse grid volume */
+    F out;               // output vector field
+    const F in;          // input vector field
+    const X xInv;        // Kahler-Dirac inverse gauge field
+    int_fastdiv X0h;     // One-half of X dimension length
+    int_fastdiv dim[4];  // full lattice dimensions
+    const int volumeCB;  // checkerboarded volume
 
-    ApplyStaggeredKDBlockArg(fineColorSpinor &out, const fineColorSpinor &in, const xInvGauge &xInv,
-                             const int *x_size_, const int *xc_size_) :
-      kernel_param(dim3(2 * in.volumeCB, 1, 1)),
+    StaggeredKDBlockArg(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &xInv) :
+      kernel_param(dim3(in.VolumeCB(), 2, 1)),
       out(out),
       in(in),
       xInv(xInv),
-      fineVolumeCB(in.volumeCB),
-      coarseVolumeCB(xInv.VolumeCB())
+      X0h(out.X()[0]/2),
+      volumeCB(in.VolumeCB())
     {
-      for (int i=0; i<QUDA_MAX_DIM; i++) {
-        x_size[i] = x_size_[i];
-        xc_size[i] = xc_size_[i];
+      if (in.V() == out.V()) errorQuda("Aliasing pointers");
+      checkOrder(out, in); // check all orders match
+      checkPrecision(out, in, xInv); // check all precisions match
+      checkLocation(out, in, xInv);
+      if (xInv.Ndim() != nDim)
+        errorQuda("Number of dimensions is not supported");
+
+      for (int i=0; i<nDim; i++) {
+        dim[i] = out.X()[i];
       }
     }
   };
 
-  template<typename Arg> struct StaggeredKDBlock {
+  template<typename Arg> struct StaggeredKDBlockApply {
+
+    using real = typename Arg::real;
+    using Vector = ColorSpinor<typename Arg::real, Arg::nColor, Arg::nSpin>;
+    using Link = Matrix<complex<real>, Arg::nColor>;
+
     const Arg &arg;
-    constexpr StaggeredKDBlock(const Arg &arg) : arg(arg) {}
+    constexpr StaggeredKDBlockApply(const Arg &arg) : arg(arg) {}
     static constexpr const char *filename() { return KERNEL_FILE; }
 
-    __device__ inline void operator()(int tid)
+    __device__ __host__ inline void operator()(int x_cb, int parity)
     {
-      // Vector type for spinor
-      using real_spinor = typename mapper<typename Arg::vFloatSpinor>::type;
-      using complex_spinor = complex<real_spinor>;
-      using Vector = ColorSpinor<real_spinor, Arg::fineColor, 1>;
-
-      // Type for gauge/compute
-      using complex = complex<typename Arg::real>;
-
-      /////////////////////////////////
-      // Figure out some identifiers //
-      /////////////////////////////////
-
-      // Decompose into factors of 16.
-      const unsigned int fast_idx = tid & 0b1111;
```

Comment:
Nothing like human readable code 😛 .  The compiler is smart enough to decipher if written in terms of `/` and `%`, please do so 😄

Reply:
This is in code that I deleted, not added, so I guess already done :)

Reply:
🤣